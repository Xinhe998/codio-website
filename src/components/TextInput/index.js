import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { IoIosWarning, IoIosCheckmarkCircle } from 'react-icons/io';
import Button from '../Button';
import './index.scss';

const TextInput = React.forwardRef(({
  type,
  placeholder,
  text,
  title,
  onChange,
  required,
  showHint,
  hintType,
  hintText,
  onFocus,
  onBlur,
  disabled,
  icon,
  showPostBtn,
  postBtnText,
  postBtnOnClick,
  readonly
}, ref) => {
  return (
    <div
      className={classNames(
        'textinput',
        required ? 'required' : null,
        showHint ? hintType || null : null,
      )}
    >
      {title ? <span className="textinput__title">{title}</span> : null}
      <input
        type={type}
        value={text}
        placeholder={placeholder}
        onChange={e => onChange(e)}
        onFocus={onFocus}
        onBlur={onBlur}
        disabled={disabled}
        readOnly={readonly}
        ref={ref}
      />
      <div className="textinput__icon">{icon}</div>
      {showPostBtn ? (
        <Button className="textinput__postBtn" type="primary" size="small" theme="blue" text={postBtnText} onClick={postBtnOnClick} />
      ) : null}
      {showHint ? (
        <span className="textinput__hint">
          <span className="hinttext">{hintText}</span>
          {hintType === 'error' ? <IoIosWarning /> : null}
          {hintType === 'ok' ? <IoIosCheckmarkCircle /> : null}
        </span>
      ) : null}
    </div>
  );
});

TextInput.propTypes = {
  title: PropTypes.string,
  type: PropTypes.string,
  text: PropTypes.string,
  placeholder: PropTypes.string,
  showHint: PropTypes.bool,
  hintType: PropTypes.string,
  hintText: PropTypes.string,
  onChange: PropTypes.func,
  required: PropTypes.bool,
  onBlur: PropTypes.func,
  onFocus: PropTypes.func,
  disabled: PropTypes.bool,
  icon: PropTypes.any,
  showPostBtn: PropTypes.bool,
  postBtnText: PropTypes.string,
  readonly: PropTypes.bool,
  postBtnOnClick: PropTypes.func,
};

TextInput.defaultProps = {
  type: 'text',
  text: '',
  onBlur: null,
  onFocus: null,
  disabled: false,
  readonly: false,
};
export default TextInput;
