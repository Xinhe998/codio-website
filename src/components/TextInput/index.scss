@import '../../../styles/main.scss';

.textinput {
  display: flex;
  flex-direction: column;
  box-sizing: border-box;
  transition: all 0.5s;
  position: relative;

  input {
    min-height: 40px;
    border-radius: 5px;
    border: solid 1px $gray;
    background-color: $white;
    outline: 0;
    max-width: 100%;
    box-sizing: border-box;
    padding: 4px 11px;
    color: rgba(0, 0, 0, 0.65);
    position: relative;
    margin: 0;
    transition: all 0.5s;
    color: black;

    &:focus {
      border: solid 1px $blue;
    }

    &:disabled {
      background-color: $gray;
      color: $dark-gray;
    }

    &::placeholder {
      color: #e4e4e4;
    }
  }

  &.required {
    .textinput__title {
      &::after {
        content: '*';
        color: $codio-pink;
        font-size: 15px;
        padding: 0 5px;
        font-weight: normal;
        font-style: normal;
        font-stretch: normal;
        line-height: 1.35;
        letter-spacing: 1;
        text-align: left;
      }
    }
  }

  &.error {
    input {
      border: solid 1px $codio-pink;
      // padding-right: 250px;
    }

    .textinput__hint {
      position: absolute;
      top: calc(100% / 2);
      right: 15px;
      font-size: 12px;
      margin: 5px 0;
      color: $codio-pink;

      .hinttext {
        vertical-align: middle;
      }

      svg {
        width: 20px;
        height: 20px;
        vertical-align: middle;
        margin-left: 10px;
      }
    }
  }

  &.ok {
    input {
      border: solid 1px $green;
      padding-right: 50px;
    }

    .textinput__hint {
      position: absolute;
      top: calc(100% / 2);
      right: 15px;
      font-size: 14px;
      margin: 3px 0;
      color: $green;

      .hinttext {
        vertical-align: middle;
      }

      svg {
        width: 20px;
        height: 20px;
        vertical-align: middle;
        margin-left: 10px;
      }
    }
  }
}

.textinput__title {
  font-size: 7px;
  font-weight: normal;
  font-style: normal;
  font-stretch: normal;
  line-height: 1.14;
  letter-spacing: normal;
  text-align: left;
  color: $dark-gray;
  margin-bottom: 7px;
}

.textinput__icon {
  position: absolute;
  top: calc(50% / 2);
  right: 5px;
  display: flex;
  justify-content: center;
  align-items: center;
  cursor: pointer;

  svg {
    width: 18px;
    height: 18px;
    color: $white;
  }
}

.textinput__postBtn {
  position: absolute;
  width: 70px !important;
  top: 0 !important;
  margin: 0 !important;
  height: 40px !important;
  border-radius: 0 5px 5px 0 !important;
  right: 0 !important;
  font-size: 12px !important;
}
