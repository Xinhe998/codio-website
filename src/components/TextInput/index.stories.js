import React, { useState } from 'react';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { DocsPage, DocsContainer } from '@storybook/addon-docs/blocks';
import StoryRouter from 'storybook-react-router';
import TextInput from './index';
import storeFs from '../../reducers';

const store = createStore(storeFs);
const stories = storiesOf('TextInput', module).addParameters({
  title: 'TextInput',
  component: TextInput,
  docs: {
    container: DocsContainer,
    page: DocsPage,
  },
});

stories.addDecorator(StoryRouter()).add('default', () => (
  <Provider store={store}>
    <TextInput
      title="E-mail"
      text=""
      showHint={false}
      hintType="ok"
      // onChange={e => setEmail(e.target.value)}
      required
    />
  </Provider>
));

stories.addDecorator(StoryRouter()).add('controlled', () => {
  const Parent = ({ children, ...props }) => {
    const [state, setState] = useState([]);
    return <div>{children(state, setState)}</div>;
  };
  return (
    <Parent>
      {(state, setState) => (
        <Provider store={store}>
          <TextInput
            title="E-mail"
            text={state}
            showHint={false}
            hintType="ok"
            onChange={e => setState(e.target.value)}
            required
          />
        </Provider>
      )}
    </Parent>
  );
});
