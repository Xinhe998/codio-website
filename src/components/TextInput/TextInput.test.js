/* eslint-disable func-names */
import 'babel-polyfill';
import React from 'react';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import { BrowserRouter as Router } from 'react-router-dom';
import { shallow, configure, render, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import TextInput from './index';

import fakeImg from '../../assets/smartmockups_jxfuqv8i.jpg';
import fakeAvatar from '../../assets/userImg.png';

import storeFs from '../../reducers';

const store = createStore(storeFs);

// 以該解析器提供給 Enzyme 做渲染 Component 的設置
configure({ adapter: new Adapter() });

describe('TextInput', function() {
  let testComp;

  beforeEach(() => {
    jest.resetModules();
    this.params = {
      type: 'text',
      placeholder: 'Test Placeholder',
      text: 'Test',
      title: 'TextInput',
      onChange: jest.fn(),
      required: false,
      showHint: false,
      hintType: 'ok',
      hintText: '',
      onFocus: jest.fn(),
      onBlur: jest.fn(),
      disabled: false,
      icon: '',
      showPostBtn: false,
      postBtnText: '',
      postBtnOnClick: jest.fn(),
      readonly: false,
    };

    this.makeSubject = () => {
      const {
        type,
        placeholder,
        text,
        title,
        onChange,
        required,
        showHint,
        hintType,
        hintText,
        onFocus,
        onBlur,
        disabled,
        icon,
        showPostBtn,
        postBtnText,
        postBtnOnClick,
        readonly,
      } = this.params;

      testComp = shallow(
        <Router>
          <Provider store={store}>
            <TextInput
              title={title}
              type={type}
              placeholder={placeholder}
              text={text}
              onChange={onChange}
              required={required}
              showHint={showHint}
              hintType={hintType}
              hintText={hintText}
              onFocus={onFocus}
              onBlur={onBlur}
              disabled={disabled}
              icon={icon}
              showPostBtn={showPostBtn}
              postBtnText={postBtnText}
              postBtnOnClick={postBtnOnClick}
              readonly={readonly}
            />
          </Provider>
        </Router>,
      );
      return testComp;
    };
  });
  describe('when it has complete data', () => {
    beforeEach(() => {
      this.subject = this.makeSubject();
    });
    it('should be same as snapshot', async () => {
      const wrapper = render(await testComp);
      expect(wrapper).toMatchSnapshot();
    });
    it('should find className', async () => {
      expect(testComp.find('.textinput'));
    });
  });
});
