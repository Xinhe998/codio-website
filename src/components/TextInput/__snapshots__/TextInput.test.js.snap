// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput when it has complete data should be same as snapshot 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "textinput",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "class": "textinput__title",
        },
        "children": Array [
          Object {
            "data": "TextInput",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
        ],
        "name": "span",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {
            "placeholder": "Test Placeholder",
            "type": "text",
            "value": "Test",
          },
          "children": Array [],
          "name": "input",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": Object {
            "attribs": Object {
              "class": "textinput__icon",
            },
            "children": Array [],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": [Circular],
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "placeholder": undefined,
            "type": undefined,
            "value": undefined,
          },
          "x-attribsPrefix": Object {
            "placeholder": undefined,
            "type": undefined,
            "value": undefined,
          },
        },
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
      Object {
        "attribs": Object {
          "placeholder": "Test Placeholder",
          "type": "text",
          "value": "Test",
        },
        "children": Array [],
        "name": "input",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {
            "class": "textinput__icon",
          },
          "children": Array [],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "class": "textinput__title",
          },
          "children": Array [
            Object {
              "data": "TextInput",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "text",
            },
          ],
          "name": "span",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "placeholder": undefined,
          "type": undefined,
          "value": undefined,
        },
        "x-attribsPrefix": Object {
          "placeholder": undefined,
          "type": undefined,
          "value": undefined,
        },
      },
      Object {
        "attribs": Object {
          "class": "textinput__icon",
        },
        "children": Array [],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "placeholder": "Test Placeholder",
            "type": "text",
            "value": "Test",
          },
          "children": Array [],
          "name": "input",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": Object {
            "attribs": Object {
              "class": "textinput__title",
            },
            "children": Array [
              Object {
                "data": "TextInput",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "text",
              },
            ],
            "name": "span",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": [Circular],
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
          "type": "tag",
          "x-attribsNamespace": Object {
            "placeholder": undefined,
            "type": undefined,
            "value": undefined,
          },
          "x-attribsPrefix": Object {
            "placeholder": undefined,
            "type": undefined,
            "value": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
